File: app/src/main/java/com/icebem/akt/activity/MainActivity.java
package com.icebem.akt.activity;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.provider.Settings;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;
import androidx.navigation.NavController;
import androidx.navigation.NavDestination;
import androidx.navigation.Navigation;
import androidx.navigation.ui.AppBarConfiguration;
import androidx.navigation.ui.NavigationUI;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.navigation.NavigationView;
import com.icebem.akt.R;
import com.icebem.akt.app.BaseApplication;
import com.icebem.akt.app.CompatOperations;
import com.icebem.akt.app.GestureActionReceiver;
import com.icebem.akt.app.PreferenceManager;
import com.icebem.akt.service.OverlayService;
import com.icebem.akt.util.AppUtil;

import java.text.SimpleDateFormat;
import java.util.Locale;

public class MainActivity extends AppCompatActivity {
    private TextView subtitle;
    private FloatingActionButton fab;
    private AppBarConfiguration barConfig;
    private NavController navController;
    private PreferenceManager manager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        setSupportActionBar(findViewById(R.id.toolbar));
        manager = PreferenceManager.getInstance(this);
        fab = findViewById(R.id.fab);
        barConfig = new AppBarConfiguration.Builder(R.id.nav_home, R.id.nav_tools, R.id.nav_settings).setDrawerLayout(findViewById(R.id.drawer_layout)).build();
        navController = Navigation.findNavController(this, R.id.nav_host_fragment);
        NavigationView navigationView = findViewById(R.id.nav_view);
        subtitle = navigationView.getHeaderView(0).findViewById(R.id.txt_header_subtitle);
        updateSubtitleTime();
        NavigationUI.setupActionBarWithNavController(this, navController, barConfig);
        NavigationUI.setupWithNavController(navigationView, navController);
        navController.addOnDestinationChangedListener((controller, destination, bundle) -> onDestinationChanged(destination));
        fab.setOnClickListener(v -> showOverlay());
        if (manager.isPro())
            fab.setOnLongClickListener(this::startGestureAction);
    }

    public void showOverlay() {
        if (CompatOperations.requireOverlayPermission(this)) {
            AlertDialog.Builder builder = new AlertDialog.Builder(this);
            builder.setTitle(R.string.state_permission_request);
            builder.setMessage(R.string.msg_permission_overlay);
            builder.setPositiveButton(R.string.permission_permit, (dialog, which) -> startActivity(new Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION)));
            builder.setNegativeButton(R.string.no_thanks, null);
            builder.create().show();
        } else
            startService(new Intent(this, OverlayService.class));
    }

    private boolean startGestureAction(View view) {
        if (((BaseApplication) getApplication()).isGestureServiceRunning()) {
            LocalBroadcastManager.getInstance(this).sendBroadcast(new Intent(GestureActionReceiver.ACTION));
        } else if (((BaseApplication) getApplication()).isGestureServiceEnabled()) {
            Toast.makeText(this, R.string.error_occurred, Toast.LENGTH_SHORT).show();
            startActivity(new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS, Uri.parse(AppUtil.URL_PACKAGE)));
        } else {
            Toast.makeText(this, R.string.info_gesture_request, Toast.LENGTH_SHORT).show();
            startActivity(new Intent(Settings.ACTION_ACCESSIBILITY_SETTINGS));
        }
        return true;
    }

    public void updateSubtitleTime() {
        subtitle.setText(new SimpleDateFormat(AppUtil.DATE_FORMAT, Locale.getDefault()).format(manager.getCheckLastTime()));
    }

    private void onDestinationChanged(NavDestination destination) {
        if (destination.getId() == R.id.nav_home) {
            if (fab.isOrWillBeHidden())
                fab.show();
        } else if (fab.isOrWillBeShown())
            fab.hide();
    }

    @Override
    public boolean onSupportNavigateUp() {
        return NavigationUI.navigateUp(navController, barConfig) || super.onSupportNavigateUp();
    }
}

File: app/src/main/java/com/icebem/akt/activity/MainActivity.kt
package com.icebem.akt.activity

import android.content.Intent
import android.os.Bundle
import android.provider.Settings
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import androidx.navigation.NavController
import androidx.navigation.NavDestination
import androidx.navigation.Navigation
import androidx.navigation.ui.AppBarConfiguration
import androidx.navigation.ui.NavigationUI
import com.google.android.material.floatingactionbutton.ExtendedFloatingActionButton
import com.google.android.material.navigation.NavigationView
import com.icebem.akt.R
import com.icebem.akt.app.BaseApplication
import com.icebem.akt.app.CompatOperations
import com.icebem.akt.app.GestureActionReceiver
import com.icebem.akt.app.PreferenceManager
import com.icebem.akt.service.OverlayService
import com.icebem.akt.util.AppUtil
import com.icebem.akt.util.DataUtil
import com.icebem.akt.util.RandomUtil
import org.json.JSONArray
import java.text.SimpleDateFormat
import java.util.*

class MainActivity : AppCompatActivity() {
    val list by lazy { arrayListOf<Int>() }
    val array by lazy {
        try {
            DataUtil.getSloganData(this)
        } catch (e: Exception) {
            JSONArray()
        }
    }
    lateinit var fab: ExtendedFloatingActionButton private set
    private lateinit var subtitle: TextView
    private lateinit var barConfig: AppBarConfiguration
    private lateinit var navController: NavController
    private lateinit var manager: PreferenceManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(findViewById(R.id.toolbar))
        manager = PreferenceManager.getInstance(this)
        barConfig = AppBarConfiguration.Builder(R.id.nav_home, R.id.nav_tools, R.id.nav_settings).setOpenableLayout(findViewById(R.id.drawer_layout)).build()
        fab = findViewById(R.id.fab)
        navController = Navigation.findNavController(this, R.id.nav_host_fragment)
        navController.addOnDestinationChangedListener { _, destination: NavDestination, _ -> onDestinationChanged(destination) }
        val navigationView = findViewById<NavigationView>(R.id.nav_view)
        subtitle = navigationView.getHeaderView(0).findViewById(R.id.txt_header_subtitle)
        NavigationUI.setupActionBarWithNavController(this, navController, barConfig)
        NavigationUI.setupWithNavController(navigationView, navController)
        fab.setOnClickListener { showOverlay() }
        if (manager.isPro) fab.setOnLongClickListener { startGestureAction() }
    }

    override fun onStart() {
        super.onStart()
        updateSubtitleTime()
    }

    val slogan: String
        get() {
            return if (array.length() > 0) {
                if (list.isEmpty()) {
                    for (i in 0 until array.length())
                        list.add(i)
                }
                val i = RandomUtil.randomIndex(list.size)
                val obj = array.getJSONObject(list[i])
                list.removeAt(i)
                getString(R.string.operator_slogan, obj.getString("slogan"), obj.getString("name"))
            } else getString(R.string.error_slogan)
        }

    fun showOverlay() {
        if (CompatOperations.requireOverlayPermission(this)) {
            AlertDialog.Builder(this).run {
                setTitle(R.string.state_permission_request)
                setMessage(R.string.msg_permission_overlay)
                setPositiveButton(R.string.permission_permit) { _, _ -> startActivity(Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION)) }
                setNegativeButton(R.string.no_thanks, null)
                create().show()
            }
        } else {
            startService(Intent(this, OverlayService::class.java))
            finishAndRemoveTask()
        }
    }

    private fun startGestureAction(): Boolean {
        when {
            (application as BaseApplication).isGestureServiceRunning -> LocalBroadcastManager.getInstance(this).sendBroadcast(Intent(GestureActionReceiver.ACTION))
            (application as BaseApplication).isGestureServiceEnabled -> {
                AlertDialog.Builder(this).run {
                    setTitle(R.string.error_occurred)
                    setMessage(R.string.error_accessibility_killed)
                    setPositiveButton(R.string.action_reinstall) { _, _ -> CompatOperations.reinstallSelf(context) }
                    setNegativeButton(android.R.string.cancel, null)
                    create().show()
                }
            }
            else -> startActivity(Intent(Settings.ACTION_ACCESSIBILITY_SETTINGS))
        }
        return true
    }

    fun updateSubtitleTime() {
        subtitle.text = SimpleDateFormat(AppUtil.DATE_FORMAT, Locale.getDefault()).format(manager.checkLastTime)
    }

    private fun onDestinationChanged(destination: NavDestination) {
        if (destination.id == R.id.nav_home) {
            if (!fab.isShown) fab.show()
        } else if (fab.isShown) fab.hide()
    }

    override fun onSupportNavigateUp(): Boolean = NavigationUI.navigateUp(navController, barConfig) || super.onSupportNavigateUp()
}